pages/api/orders/AddOrder.js

```diff
@@ -0,0 +1,125 @@
+import {validateFunc} from '../validate';
+import { google } from 'googleapis'; //npm install googleapis
+import updateGoogleSheets from './UpdateGoogleSheets';
+import firebase from "firebase/app";
+import useSWR from 'swr';
+import { resolveHref } from 'next/dist/next-server/lib/router/router';
+export const config = { // https://nextjs.org/docs/api-routes/api-middlewares
+  api: {
+    bodyParser: true,
+  },
+} 
+
+// when you call addorder.js, it returns a promise 
+function requireParams(body, res) {
+    /* require elements: First name, last name, email address, address, calID, 
+    delivery_date, order_timestamp, items in the order (orderschema as of now)
+    items are represented as an object with barcode as key and quantity (count) as value*/
+
+    if (!body.firstName || !body.lastName || !body.address ||
+        !body.emailAddress || !body.calID || !body.items || !body.deliveryDate) {
+          res.json({error: "missing firstName||lastName||EmailAddress||Address||calID||items||deliveryDate\
+             in request"}); 
+            res.status(400);
+            return false;
+        }
+    //require order items object with at least one entry (order array)
+    if (body.items.length <= 0) {
+        res.json({error: "missing order item"}); 
+        res.status(400);
+        return false;
+    } 
+    return true;
+}
+
+
+function updateInventory(items) { //updates firebase
+  return new Promise((resolve, reject) => {
+    let inventory = fetch(process.env.GET_ALL_ITEMS);//returns inventory as object 
+    inventory.then((value) => {
+      value.json().then((inventoryJson) => {
+         const inventoryUpdates = {}
+         for (let item in items) {
+  
+           if (inventoryJson[item]['count'] >= items[item]) { //if we have enough in inventory for order
+            inventoryUpdates['/inventory/' + item + "/count"] = firebase.database.ServerValue.increment(-1 * items[item]);
+           }
+           else {
+             console.log("Sorry, requested count for " + inventoryJson[item]["itemName"] + " exceeds inventory");
+             return reject("Quantity exceeded"); 
+           }
+         }
+         firebase.database().ref().update(inventoryUpdates).then(() => {
+           return resolve("Inventory updated");
+         })
+         .catch((error) => {
+           console.log("error updating to firebase: ", error);
+           return reject();
+         }) 
+      })  
+    }) 
+  })
+}
+
+function addOrder(firstName, lastName, address, emailAddress, calID, items, deliveryDate) {
+      //add order data to google sheets
+    const target = ['https://www.googleapis.com/auth/spreadsheets'];
+    const jwt = new google.auth.JWT(
+      process.env.GOOGLE_SHEETS_CLIENT_EMAIL,
+      null,
+      (process.env.GOOGLE_SHEETS_PRIVATE_KEY || '').replace(/\\n/g, '\n'),
```

Comment:
what is `replace(/\\n/g, '\n'),` for?

Reply:
I got this template from here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace 
I believe it replaces '\\n' with '\n', as there's a 'no start line' error when I remove it.